contracts/signature-bridge/src/contract.rs

```diff
@@ -0,0 +1,276 @@
+
+    let msgs: Vec<CosmosMsg> = vec![CosmosMsg::Wasm(WasmMsg::Execute {
+        contract_addr: msg.handler_addr,
+        funds: vec![],
+        msg: to_binary(&ExecutorExecMsg::SetResource {
+            resource_id: msg.resource_id,
+            contract_addr: msg.execution_context_addr,
+        })
+        .unwrap(),
+    })];
+
+    Ok(Response::new()
+        .add_messages(msgs)
+        .add_attributes(vec![attr("method", "admin_set_resource_with_sig")]))
+}
+
+fn exec_proposal_with_signature(
+    deps: DepsMut,
+    _info: MessageInfo,
+    msg: ExecProposalWithSigMsg,
+) -> Result<Response, ContractError> {
+    let state = STATE.load(deps.storage)?;
+
+    // Validations
+    if !signed_by_governor(&msg.data, &msg.sig, state.governor.as_str())? {
+        return Err(ContractError::Std(StdError::GenericErr {
+            msg: "Invalid sig from governor".to_string(),
+        }));
+    }
+
+    // Parse resourceID from the data
+    let resource_id_bytes = &msg.data[0..32];
+    let resource_id = element_encoder(resource_id_bytes);
+
+    // Parse chain ID + chain type from the resource ID
+    let execution_chain_id_type: u64 = get_chain_id_type(&resource_id_bytes[26..32]);
+
+    // Verify current chain matches chain ID from resource ID
+    if compute_chain_id_type(MOCK_CHAIN_ID, &COSMOS_CHAIN_TYPE) != execution_chain_id_type {
+        return Err(ContractError::Std(StdError::GenericErr {
+            msg: "executing on wrong chain".to_string(),
+        }));
+    }
+
+    // Handle the main business
+    let handler_addr = RESID2HANDLERADDR
+        .load(deps.storage, &resource_id)?
+        .to_string();
+    let msgs: Vec<CosmosMsg> = vec![CosmosMsg::Wasm(WasmMsg::Execute {
+        contract_addr: handler_addr,
+        funds: vec![],
+        msg: to_binary(&ExecutorExecMsg::ExecuteProposal {
+            resource_id,
+            data: msg.data,
+        })
+        .unwrap(),
+    })];
+
+    Ok(Response::new()
+        .add_messages(msgs)
+        .add_attributes(vec![attr("method", "execute_proposal_with_sig")]))
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn query(deps: Deps, _env: Env, msg: QueryMsg) -> StdResult<Binary> {
+    match msg {
+        QueryMsg::GetState {} => to_binary(&get_state(deps)?),
+    }
+}
+
+fn get_state(deps: Deps) -> StdResult<StateResponse> {
+    let state = STATE.load(deps.storage)?;
+    Ok(StateResponse {
+        governor: state.governor.to_string(),
+        proposal_nonce: state.proposal_nonce,
+    })
+}
+
+// Verifying signature of governor over some datahash
+fn signed_by_governor(data: &[u8], sig: &[u8], governor: &str) -> Result<bool, ContractError> {
+    let hashed_data = Keccak256::hash(data).map_err(|_| ContractError::HashError)?;
+    let message = Message::parse(&hashed_data);
+    let sig = Signature::parse_standard_slice(sig).map_err(|_| ContractError::HashError)?;
+    let recv_id = RecoveryId::parse(0_u8).map_err(|_| ContractError::HashError)?;
```

Comment:
This looks incorrect. Please refer to https://docs.rs/sp-io/4.0.0/src/sp_io/lib.rs.html#770-793.

The `recv_id` should come from the signature data itself.

Reply:
Replaced the `libsecp256` lib with the cosmwasm-builtin `secp256k1_verify` util.
https://github.com/webb-tools/protocol-cosmwasm/pull/37/commits/2b845e7ee8119bba81d3ff3b693de047f55ba077